module/zcommon/zfs_fletcher.c

```diff
@@ -300,33 +300,47 @@ fletcher_2_byteswap(const void *buf, uint64_t size,
 	(void) fletcher_2_incremental_byteswap((void *) buf, size, zcp);
 }
 
-ZFS_NO_SANITIZE_UNDEFINED
+#if defined(__GNUC__) && !defined(__clang__)
+__attribute__((optimize("no-tree-vectorize")))
+#elif defined(__clang__)
+__attribute__((optnone))
+#endif
 static void
 fletcher_4_scalar_init(fletcher_4_ctx_t *ctx)
 {
-	ZIO_SET_CHECKSUM(&ctx->scalar, 0, 0, 0, 0);
+	ZIO_SET_CHECKSUM((zio_cksum_t *)ctx, 0, 0, 0, 0);
```

Comment:
To make sure the `.scaler` field always remains the first member of the structure let's add a static assert to enforce that.  Something like this after the structure definition. 

```c
_Static_assert(offsetof (struct fletcher_4_ctx_t, scalar) == 0,
     "zio_cksum_t scaler field must be first member");
```

This would also be a good place to leave a comment about why this is important.  Since it's not at all clear from a causal (or even careful!) reading why it would matter.

Reply:
I'll happily go add the static assert, but...

> This is definitely better that the current situation. Alternately, I wonder if we could force the correct alignment with the `aligned` attribute so it could be safely vectorized.

You could, yes - dropping the no-vectorize defines, the compiler correctly emits unaligned accesses with this change, on my testbed, compared to the aligned writes that were dying before.

I'd argue we absolutely don't want at least _this one_ to vectorize, though, since we explicitly use this one in cases where we have remaining bytes that didn't neatly align to the vectorized implementations' requirements. (The superscalar ones do get a speedup if you vectorize them, unsurprisingly, though not to the same degree as if you actually ran the SSE or AVX ones.)

Reply:
Oh, and for forcing the alignment, we could try another union, but I don't think it's worth doing it here. Dropping the UB and fixing the bug is more important for now.